Sources/FoundationEssentials/AttributedString/AttributedString+IndexTracking.swift

```diff
@@ -0,0 +1,168 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#if FOUNDATION_FRAMEWORK
+@_spi(Unstable) internal import CollectionsInternal
+#elseif canImport(_RopeModule)
+internal import _RopeModule
+#elseif canImport(_FoundationCollections)
+internal import _FoundationCollections
+#endif
+
+// MARK: - Internal Index Updating
+
+extension AttributedString.Guts {
+    func _prepareTrackedIndicesUpdate(mutationRange: Range<BigString.Index>) {
+        // Move any range endpoints inside of the mutation range to outside of the mutation range since a range should never end up splitting a mutation
+        for idx in 0 ..< trackedRanges.count {
+            let lowerBoundWithinMutation = trackedRanges[idx].lowerBound > mutationRange.lowerBound && trackedRanges[idx].lowerBound < mutationRange.upperBound
+            let upperBoundWithinMutation = trackedRanges[idx].upperBound > mutationRange.lowerBound && trackedRanges[idx].upperBound < mutationRange.upperBound
+            switch (lowerBoundWithinMutation, upperBoundWithinMutation) {
+            case (true, true):
+                // Range is fully within mutation, collapse it to the start of the mutation
+                trackedRanges[idx] = Range(uncheckedBounds: (mutationRange.lowerBound, mutationRange.lowerBound))
+            case (true, false):
+                // Range starts within mutation but extends beyond mutation - remove portion within mutation
+                trackedRanges[idx] = Range(uncheckedBounds: (mutationRange.upperBound, trackedRanges[idx].upperBound))
+            case (false, true):
+                // Range starts before mutation but extends into mutation - remove portion within mutation
+                trackedRanges[idx] = Range(uncheckedBounds: (trackedRanges[idx].lowerBound, mutationRange.lowerBound))
+            case (false, false):
+                // Neither endpoint of range is within mutation, leave as-is
+                break
+            }
+        }
+    }
+    
+    func _finalizeTrackedIndicesUpdate(mutationStartOffset: Int, isInsertion: Bool, utf8LengthDelta: Int) {
+        // Update indices to point to the correct offsets based on the mutation deltas
+        for idx in 0 ..< trackedRanges.count {
+            var lowerBound = trackedRanges[idx].lowerBound
+            var upperBound = trackedRanges[idx].upperBound
+            
+            // Shift the lower bound if either:
+            //      A) The lower bound is greater than the start of the mutation (meaning it must be after the mutation due to the prepare step)
+            //      B) The lower bound is equal to the start of the mutation, but the mutation is an insertion (meaning the text is inserted before the start offset)
+            if lowerBound.utf8Offset > mutationStartOffset || (lowerBound.utf8Offset == mutationStartOffset && isInsertion), utf8LengthDelta != 0 {
```

Comment:
The comment suggests that you want to shift the lower bound if either of these condition satisfies, but it doesn't mention how `utf8LengthDelta` affects this.  Just checking that we're not missing this just because `utf8LengthDelta == 0`

Reply:
Ah right the comment doesn't mention that - that check is just so that if the mutation doesn't actually change the length of the string (i.e. "ABC" was replaced with "DEF") then there's no need to re-calculate offsets, we just do the same thing we would have done if the index was before the mutation. I can add a note to the comment here to make that a bit clearer